### PR TITLE
Handle Pester Describe block strings with single quotes inside it

### LINKED ISSUE #1725: Single quotes inside Pester test names prevent tests from running
## Issue Description ##

The "Run Tests" or "Debug Tests" option appearing above Pester Describe blocks does not work when the string contains a single quote. vscode-powershell appears to be neglecting to escape them when issuing a shell command to run the tests.

```
Describe "This test works" {
    It "follows the basic principles of logic" {
        $true | should be $true
    }
}

Describe "This test's broken" {
    It "follows the basic principles of logic" {
        $true | should be $true
    }
}
```

Working test yields:

```
Describing This test works
 [+] follows the basic principles of logic 56ms
Tests completed in 56ms
Passed: 1 Failed: 0 Skipped: 0 Pending: 0 Inconclusive: 0
```

And broken test:

```
The string is missing the terminator: '.
At line:0 char:0
```

I'm still using version 3.4.0 of Pester, if that matters. The tests run fine when I call `Invoke-Pester` in the console.

### Visual Studio Code ###

| Name | Version |
| --- | --- |
| Operating System | Windows_NT x64 10.0.15063 |
| VSCode | 1.30.2|
| PowerShell Extension Version | 1.10.2 |

### PowerShell Information ###

|Name|Value|
|---|---|
|PSVersion|5.1.15063.1563|
|PSEdition|Desktop|
|PSCompatibleVersions|1.0 2.0 3.0 4.0 5.0 5.1.15063.1563|
|BuildVersion|10.0.15063.1563|
|CLRVersion|4.0.30319.42000|
|WSManStackVersion|3.0|
|PSRemotingProtocolVersion|2.3|
|SerializationVersion|1.1.0.1|

### Visual Studio Code Extensions ###

<details><summary>Visual Studio Code Extensions(Click to Expand)</summary>

|Extension|Author|Version|
|---|---|---|
|PowerShell|ms-vscode|1.10.2|

(All other extensions manually disabled to confirm no interference)

### PR DESCRIPTION
**Pull request by bergmeister**
## PR Summary

Fixes #1725

Although PR #1713 will fix any problems of running Pester tests with special describe block strings when the latest version of Pester (>4.6.0) is used, this PR fixes the behaviour when an older version of Pester is installed to allow running tests with describe blocks that contain a single quote inside them. I also checked that this does not break running describe blocks with single quotes strings or single quotes inside single quotes. I also tested multiple single quotes inside the string.
image

## PR Checklist

Note: Tick the boxes below that apply to this pull request by putting an `x` between the square brackets.
Please mark anything not applicable to this PR `NA`.

- [x] PR has a meaningful title
- [x] Summarized changes
- [ ] PR has tests
- [x] This PR is ready to merge and is not work in progress
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready

### CODE REVIEW

**Comment by rkeithhill** [comment]
I wonder if the fix for this should be in PSES where we get the TestName before we send it back to the extension.  We also have a method there that does the single quote escaping.  It is called PowerShellContext.QuoteEscapeString().  Since this method provides the outer single quotes, we'd need to remove those in the extension.  I can see pluses/minuses to both approaches but I do like that PSES in this case declares the string is a single quoted string rather than leaving the interpretation up the client.

**Comment by bergmeister** [comment]
The advantage of doing things in PSES would be performance to me but for this case I see it as negligible. I found the `QuoteEscapeString` quite complicated and for what it is doing and I'm not even sure if using a StringBuilder has a better performance than this simple code
```csharp
var testName = pesterSymbol.TestName;
if (testName.Contains("'"))
{
    testName = testName.Replace("'", "''");
}
```
If we decide to do it in PSES we'd still need a PR in the extension itself to remove the single quote outside of the test name that is currently hard-coded in it.

File: src/features/PesterTests.ts [hunk 1]
```diff
@@ -72,6 +72,10 @@ export class PesterTestsFeature implements IFeature {
 
         if (describeBlockName) {
             launchConfig.args.push("-TestName");
+            // Escape single quotes inside double quotes
+            if (describeBlockName.includes("'")) {
+                describeBlockName = describeBlockName.replace("'", "''");
```
[the remark this answers is not in the document]
Comment (rkeithhill) [reply src/features/PesterTests.ts]: Ya, the JS replace function bites me from time to time - coming from .NET where it works like it should.  :-)

[the remark this answers is not in the document]
Comment (bergmeister) [reply src/features/PesterTests.ts]: Ok, thanks, I did not know that. I played around with it a bit in the console and it seems that replace basically breaks when there is more than 1 occurrence to be replaced. Yes, coming from a .Net world as well. Using `.replace(/'/g, "''")` seems to do the job

**Comment by TylerLeonhardt** [comment]
Good point @rkeithhill I'm curious about that too

**Comment by rkeithhill** [comment]
OK, couldn't remember if we were grabbing the expression's Value or Extent.Text property.  Looks good then.  

BTW did that t-shirt but imagine how much worse it would be if we were using JavaScript directly instead of TypeScript.

**Pull request merged**